hazelcast/src/main/java/com/hazelcast/internal/util/collection/ImmutableLazyEntry.java

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright (c) 2008-2023, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.internal.util.collection;
+
+import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.core.HazelcastInstanceAware;
+import com.hazelcast.internal.nio.IOUtil;
+import com.hazelcast.internal.serialization.Data;
+import com.hazelcast.internal.serialization.SerializationService;
+import com.hazelcast.nio.ObjectDataInput;
+import com.hazelcast.nio.ObjectDataOutput;
+import com.hazelcast.nio.serialization.IdentifiedDataSerializable;
+import com.hazelcast.spi.impl.SerializationServiceSupport;
+
+import java.io.IOException;
+import java.util.Map.Entry;
+import java.util.Objects;
+
+/**
+ * Implementation of {@link Entry} which lazily deserializes keyData and valueData on access.
+ * This class is not thread-safe.
+ * <p>
+ * @param <K> key
+ * @param <V> value
+ */
+public final class ImmutableLazyEntry<K, V> implements Entry<K, V>, HazelcastInstanceAware, IdentifiedDataSerializable {
+
+    private Data keyData;
+    private Data valueData;
+
+    private transient K key;
+    private transient V value;
+    private transient SerializationService serializationService;
+
+    ImmutableLazyEntry() {
+    }
+
+    public ImmutableLazyEntry(Data keyData, Data valueData, SerializationService serializationService) {
+        this.keyData = keyData;
+        this.valueData = valueData;
+        this.serializationService = serializationService;
```

Comment:
Does this mean we will set serializationService twice? I think HazelcastInstanceAware classes have their setHazelcastInstance method called some point in time. Is HazelcastInstanceAware for deserialized ImmutableLazyEntrys ?

Reply:
Yes.

This constructor is for manual creation of the Entry, e.g. in `MapEventJournalFunctions`.
`HazelcastInstanceAware` is for de-serialized Entries.